File: ios/Packages/MemoryMap/Sources/Domain/RegionSearch.swift
import CoreModel

/// 지역 검색 순서. 안드로이드 `RegionSearch` 와 같아야 두 앱이 같은 결과를 냅니다.
///
/// 1. 이름이 정확히 같은 것 → 2. 앞부분이 맞는 것 → 3. 이름에 들어 있는 것 → 4. 상위 이름에 들어 있는 것
/// 같은 등급이면 국내 시군구 → 나라 → 해외 시도 → 장소 순서입니다.
public enum RegionSearch {

    public static func rank(query: String, regions: [Region], limit: Int = 30) -> [Region] {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty else { return [] }

        let scored: [(Region, Int)] = regions.compactMap { region in
            guard let s = score(query: q, region: region) else { return nil }
            return (region, s)
        }

        return scored
            .sorted { lhs, rhs in
                if lhs.1 != rhs.1 { return lhs.1 < rhs.1 }
                let lk = kindOrder(lhs.0.code.kind), rk = kindOrder(rhs.0.code.kind)
                if lk != rk { return lk < rk }
                if lhs.0.name.count != rhs.0.name.count { return lhs.0.name.count < rhs.0.name.count }
                return lhs.0.displayName < rhs.0.displayName
            }
            .prefix(limit)
            .map(\.0)
    }

    static func score(query: String, region: Region) -> Int? {
        let name = region.name.lowercased()
        let q = query.lowercased()
        if name == q { return 0 }
        if name.hasPrefix(q) { return 1 }
        if name.contains(q) { return 2 }
        if let parent = region.parentName?.lowercased(), parent.contains(q) { return 3 }
        return nil
    }

    static func kindOrder(_ kind: RegionKind) -> Int {
        switch kind {
        case .koreanDistrict: return 0
        case .country: return 1
        case .subdivision: return 2
        case .place: return 3
        }
    }
}

File: android/domain/src/main/kotlin/kr/jjaguk/domain/RegionSearch.kt
package kr.jjaguk.domain

import kr.jjaguk.core.model.Region
import kr.jjaguk.core.model.RegionKind

/**
 * 지역 검색. 웹과 같은 느낌이어야 해서 순서 규칙을 여기 고정합니다.
 *
 * 1. 이름이 정확히 같은 것
 * 2. 이름이 검색어로 시작하는 것
 * 3. 이름에 검색어가 들어 있는 것
 * 4. 상위 이름(서울, 일본 …)에 들어 있는 것
 *
 * 같은 등급이면 **국내 시군구 → 나라 → 해외 시도 → 장소** 순서입니다.
 * 한국에서 쓰는 앱이라 "중구" 를 치면 서울 중구가 먼저 나와야 합니다.
 */
object RegionSearch {

    fun rank(query: String, regions: List<Region>, limit: Int = 30): List<Region> {
        val q = query.trim()
        if (q.isEmpty()) return emptyList()

        return regions
            .mapNotNull { region -> score(q, region)?.let { region to it } }
            .sortedWith(
                compareBy(
                    { (_, score) -> score },
                    { (region, _) -> kindOrder(region.code.kind) },
                    { (region, _) -> region.name.length },
                    { (region, _) -> region.displayName },
                )
            )
            .take(limit)
            .map { it.first }
    }

    private fun score(query: String, region: Region): Int? {
        val name = region.name
        return when {
            name.equals(query, ignoreCase = true) -> 0
            name.startsWith(query, ignoreCase = true) -> 1
            name.contains(query, ignoreCase = true) -> 2
            region.parentName?.contains(query, ignoreCase = true) == true -> 3
            else -> null
        }
    }

    private fun kindOrder(kind: RegionKind): Int = when (kind) {
        RegionKind.KoreanDistrict -> 0
        RegionKind.Country -> 1
        RegionKind.Subdivision -> 2
        RegionKind.Place -> 3
    }
}
